[WIP] ValidationResult

bakura10 opened the pull request.

Maks3w commented:
This is close. Can you just do a PR with only this files:

src/Result/ValidationResult.php
src/Result/ValidationResultInterface.php
test/Result/ValidationResultTest.php

Maks3w commented:
Result objects should be inmutable. I don't like the idea of the merge method.

bakura10 commented:
This is needed for the ValidatorChain. If you have a better idea is be happy to hear!

Envoyé de mon iPhone

> Le 19 juil. 2015 à 18:02, Maks3w notifications@github.com a écrit :
> 
> Result objects should be inmutable. I don't like the idea of the merge method.
> 
> —
> Reply to this email directly or view it on GitHub.

Maks3w commented:
ValidatorChain has the responsible of merge multiple results, not the result object.

bakura10 commented:
Make sense. I'll try with an implementation 

Envoyé de mon iPhone

> Le 19 juil. 2015 à 18:06, Maks3w notifications@github.com a écrit :
> 
> ValidatorChain it's the responsible of merge multiple results, not the result object.
> 
> —
> Reply to this email directly or view it on GitHub.

Maks3w commented:
getValidator, getContext and getData is the vector of elements for obtain the result object. If include one of them then include all.

Maks3w commented:
I think the interpolation and translation of messages should be done in a different class (MVC).

Result object is the inmutable object. Presentation layer is out of the scope.

Maks3w commented:
Should the result object implement __toString()?

Maks3w commented on src/Result/ValidationResultInterface.php:
```diff
@@ -0,0 +1,64 @@
+    public function merge(ValidationResultInterface $validationResult);
+    
+    /**
+     * Get the data
```
Explain better this. What data?

Maks3w commented on src/Result/ValidationResultInterface.php:
```diff
@@ -0,0 +1,64 @@
+    /**
+     * Get the raw error messages (in those messages, variables are not injected, useful for translation)
+     *
+     * @return array
```
array could be replaced by `string[]`

Maks3w commented:
I have some concerns too about to have two independent arrays for messages and variables.

Each message use a set or subset of variables. I think get*Messages should return a touple with the message string and the set of variables (bidimensional array).

bakura10 commented:
That's how it was done until now. I'd like to remove the messageVariables but unfortunately it's a bit harder than that. The problem is that getMessages should return ready to consume messages. By passing the variables you are basiclaly leaving the user with doing the interpolation? That's definitely not very user friendly.

bakura10 commented:
Why only the validation result @Maks3w ? The ValidatorChain has its place too, it's really a differnet validator in its behaviour and show a complex usage of ValidatorResult so I think it's uesful to have it herE.

Maks3w commented:
Should we have a getErrorCodes method? I found very useful a fixed error code for to develop adhoc UI alternatives based on the error code

Maks3w commented:
@bakura10 I don't understand your last message What was the observation?

bakura10 commented:
In your first message you told me to only keep ValidationResult and ValidationResultInterface as part of this PR, but ValidatorChain has its place too.

Other feedbacks:
- What did you mean by implementing __toString? A validation result can have multiple error messages, how should __toString work in that case?
- getErrorCodes is indeed a nice idea. The only issue I have is when using ValidatorChain. Nothing prevent multiple validators to have the same error key name. In that case, this would mean you'd loose some information.

bakura10 commented:
I'm wonder if we should not introduce a new object "ValidationResultSet" that would wrpa multiple ValidationResult. I'm going to try this ;)

Maks3w commented:
I'm just focusing in the interface the rest of changes are unrelated or non relevant in this step

toString could be an implode of messages

error codes could be namespaced for avoid  collisions 'foo.0x1 or foo.wronIP'

Current result object is already a set of messages I don't see the point of to have a different object

bakura10 commented:
I've updadted the code:
- ValidationResult is now immutable, merging is handled in ValidationChain.
- Added a new `getContext` method to ValidationResult.
- Added a `__toString` method in ValidationResult to implode each error message.

For your feedback regarding error code, I'm not sure how should we decide for the prefix. The class name ? That would be the only way to make sure we have no conflict, and I'm not sure something like "\Zend\Validator\Between.tooShort" would be really easy to consume, so for now I'm not integrating it ;).

Maks3w commented:
why data and context have to be part of the result?

bakura10 commented:
No particular reason, but I think it may be useful to have the origianl data that was submitted as part of the validation (maybe for rendering purpose?).

Maks3w commented:
for rendering you have the getVariables method. 

I think should  be removed

bakura10 commented:
Well, getVariables is a bit different. Imagine that you want to display the message "You have entered "foo", here are the error messages: xxx". The "getVariables" methods does not reutrn the original value that was submitted.

But I'm not against simplifying the interface and removing data and context from the validation result.

Maks3w commented:
The result have to answer two questions:
- Is it the result valid? 
- If not, Why it's not valid? 

If does not answer thesr questions then should be removed. 

May we could redact the description as

```
* This interface provide methods for to know the result of a validation and the reasons when the result is not valid
*
* This interface resolve the following questions:
*    * Is the result valid? 
*    * If not, Why is not valid? 
```

bakura10 commented:
Updated!

Maks3w commented:
Good!

Maks3w commented:
I'll recap the things under discussion:
- [ ] Simplify getRawMessages, getMessages and getMessagesVariables.
- [ ] Add a getErrorCodes method.
- [ ] Study if __toString is necessary
- [ ] Study if enforce namespace error codes.

New things under discussion:
- Enforcement of Serializable and JsonSerializable.
  IMO this should be implemented when required.
  While the interface enforce a common API of methods, serialized output does not enforce a common format of representation.
  This is, it's not possible ensure interoperability with 2 Json serialized results. JsonSerializable is schema free and may to have different schema each one.
- Name `Zend\Validator\Result\ValidatorResultInterface` as `Zend\Validator\ResultInterface`

Maks3w commented:
About the messages simplification I start to like the idea of provide a callback

```
getMessage($renderCallback) {
   $messages = array_map(
       $this->messages,
       function($msg) { 
           return vprintf(
                $renderCallback($message, $translationDomain),
                $variables
           )
       }
   );
   return $messages;
}
```

_note we now need to provide a translation domain_

bakura10 commented:
We should keep translation out of this. The translation will be handled by zend\i18n that can accept a validationResult

Envoyé de mon iPhone

> Le 19 juil. 2015 à 21:59, Maks3w notifications@github.com a écrit :
> 
> About the messages simplification I start to like the idea of provide a callback
> 
> getMessage($renderCallback) {
>    $messages = array_map($this->messages, function($msg) {
>    return $renderCallback($message, $variable, $translationDomain);
> }
>    return $messages;
> }
> note we now need to provide a translation domain
> 
> —
> Reply to this email directly or view it on GitHub.

Maks3w commented:
we can't, if we provide messages with placeholders is for translation isn't?

Maks3w commented:
If it was not clear, $renderCallback could be a TranslatorInterface method.

Maks3w commented:
I have edited the callback example because variables is not required when translate

Maks3w commented:
Sorry I started to be confused :) My previous example was good.

This is an example with interpolation _and_ translation

```
getMessages($renderCallback) {
  $messages = array_map(
    $this->messages,
    function($msg) { 
      return $renderCallback($message, $variables, $translationDomain);
    }
  );

  return $messages;
}

$interpolateAndTranslate = function($message, $variables, $translationDomain) {
  return vprintf(
    Zend\I18n\TranslatorInterface::translate($message, $translationDomain),
    $variables
  );
};

print_r(FooValidator->validate(data, context)->getMessages($interpolateAndTranslate));
```

malukenho commented on src/AbstractValidator.php:
```diff
@@ -0,0 +1,107 @@
+        // We cast to array to keep the same logic, as some validator may throw
+        // two error messages
+        $errorKeys          = (array) $errorKeys;
+        $errorMessages = [];
```
align

bakura10 commented:
I'm not a big fan of your approach @Maks3w , as it will forces people to define a default callback for rendering. I much prefer an approach like this:

``` php
$result = $validator->validate('foo');
$translatedResult = $translator->translateValidationResult($result);
```

In all cases either i18n will be coupled to Validator at some point, or Validator coupled to i18n, but as there are greater chances for Validator to be used, I'd prefer to introduce the dependency to the less used component.
